### PR TITLE
Add Mutation Testing Blog Post

### PR DESCRIPTION
**Pull request by Lms24**
This PR adds a blog post entry about my hackweek project: Mutation Testing in the JS SDK repo. 

It's quite lengthy so I'd appreciate a review. Might need to cut it down a bit...

### CODE REVIEW

File: data/blog/js-mutation-testing-our-sdks.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+- A good portion of created mutants survived. In Stryker-terms, this specifically means that at least one test covered the mutant but did not detect it. A fair amount could be attributed to edge case code paths, where we'd log warnings (something we just didn't test on too much) or early return in a function. The importance varied from "We should definitely cover this" to "sure, would be nice to cover but not necessary". Some other created mutants were simply a bit unrealistic, for example when an empty string was replaced with a `"Stryker was here"` constant. Some of these cases would have been caught by TypeScript at build time, due to us using string literal types quite a bit.
+- In addition to survived mutants, mutants in an almost equal number were not covered by a any unit test. We were a bit surprised about this number but it makes sense (bear with me!). It's important to keep in mind that creating mutants everywhere in our code base is realistic, given bugs will sneak the easiest into uncovered code.
```
Comment (andreiborza) [inline data/blog/js-mutation-testing-our-sdks.md:107]:
```suggestion
- In addition to survived mutants, mutants in an almost equal number were not covered by any unit test. We were a bit surprised about this number but it makes sense (bear with me!). It's important to keep in mind that creating mutants everywhere in our code base is realistic, given bugs will sneak the easiest into uncovered code.
```

File: data/blog/js-mutation-testing-our-sdks.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+More recently, we saw the light of _incremental_ mutation testing, where mutants would only be created in _changed_ code files vs in the entire code base.
+The idea is to establish a base line from a previous mutation test, compute the diff or file changes since then and start an incremental MT run on the changed files.
+This significantly decreases MT runtime but is limited in accuracy, depending on how well tools are able to determine what actually changed.
+
+## JavaScript Tooling
+
+When it comes to Mutation Testing tooling in JavaScript, the by far most popular option is [Stryker](https://stryker-mutator.io/).
```
Comment (nicohrubec) [inline data/blog/js-mutation-testing-our-sdks.md:65]:
```suggestion
When it comes to tooling in JavaScript for Mutation Testing, the by far most popular option is [Stryker](https://stryker-mutator.io/).
```

File: data/blog/js-mutation-testing-our-sdks.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+- Every package has unit tests that test individual, package-specific behavior. On this test level, we want to cover individual package exports, edge cases as well as more complicated paths, of course with a sprinkle of general purpose tests. Basically, your typical unit test setup. We mostly use [Vitest](https://vitest.dev/) and [Jest](https://jestjs.io/) (we're slowly moving away from it but [it's complicated](https://github.com/getsentry/sentry-javascript/pull/13458)).
+- On the other end of the spectrum, we created an army of [End-to-End test applications](https://github.com/getsentry/sentry-javascript/tree/develop/dev-packages/e2e-tests/test-applications). These are small standalone apps build with various frameworks in which we test our actual NPM packages. Testing here is done on a higher level, as we mostly only check the resulting payloads of errors or other events our SDKs send to Sentry. But they catch a surprisingly high number of bugs and have proven themselves worthy more than once. For E2E tests, we rely on [Playwright](https://playwright.dev/) which we like a lot.
```
Comment (nicohrubec) [inline data/blog/js-mutation-testing-our-sdks.md:81]:
```suggestion
On the other end of the spectrum, we created an army of [End-to-End test applications](https://github.com/getsentry/sentry-javascript/tree/develop/dev-packages/e2e-tests/test-applications). These are small standalone apps build with various frameworks in which we test our actual NPM packages. Testing here is done on a higher level, as we mostly only check the resulting payloads of errors or other events our SDKs send to Sentry, but they catch a surprisingly high number of bugs and have proven themselves worthy more than once. For E2E tests, we rely on [Playwright](https://playwright.dev/), which we like a lot.
```

**Review by andreiborza — COMMENTED** [review]
Some small fixes/notes. Love the post, great job!

**Review by nicohrubec — COMMENTED** [review]
Just a few suggestions. Nice post!

File: data/blog/js-mutation-testing-our-sdks.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+summary: "Let's dive into mutation testing. In this post we talk about how mutation testing works, the state of it in JavaScript and we share our results of applying it to our JavaScript SDK repository"
+images: [/images/js-mutation-testing-our-sdks/hero.jpg]
+layout: PostLayout
+canonicalUrl: mutation-testing-js-sdks
+authors: [lukasstracke]
+---
+
+Have you ever thought about testing your tests? Check how good they _really_ are at catching bugs? Welcome to the realms of test suite evaluation!
+Mutation testing is considered the gold standard when it comes to evaluating the fault detection ability (I'm gonna stop with academic terms in a sec!) of your tests.
+In this article, we'll explore what Mutation Testing is, why it's a rather niche technique and what we learnt from using it on our JavaScript SDK codebase.
```
[the remark this answers is not in the document]
Comment (Lms24) [reply data/blog/js-mutation-testing-our-sdks.md:15]: Yeah, I'll settle on "mutation testing". Thanks!

File: data/blog/js-mutation-testing-our-sdks.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+     return num % 2 !== 0
+   }
+   ```
+   The resulting code is called a "mutant".
+2. Run your tests against the mutant.
+   - If at least one of your test fails, crashes or times out, the mutant was _killed_.
+   - If all your tests still pass without a crash or timeout, the mutant _survived_.
+3. Repeat 1 and 2 until all possible mutant operators have been applied to your codebase or you reached a pre-defined max number of mutants.
+
+After this virtual bloodbath, you can calculate a _Mutation Score_ by dividing the number of all created mutants by the number of killed mutants. The score now tells you how likely it is that your tests would catch an actual bug.
```
[the remark this answers is not in the document]
Comment (Lms24) [reply data/blog/js-mutation-testing-our-sdks.md:42]: lol yes, good catch, thanks! 😅

File: data/blog/js-mutation-testing-our-sdks.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+More recently, we saw the light of _incremental_ mutation testing, where mutants would only be created in _changed_ code files vs in the entire code base.
+The idea is to establish a base line from a previous mutation test, compute the diff or file changes since then and start an incremental MT run on the changed files.
+This significantly decreases MT runtime but is limited in accuracy, depending on how well tools are able to determine what actually changed.
+
+## JavaScript Tooling
+
+When it comes to Mutation Testing tooling in JavaScript, the by far most popular option is [Stryker](https://stryker-mutator.io/).
+Stryker is available for various language but for JS specifically, it shines with support for all major test frameworks (with one unfortunate exception), support for per-test mutation coverage, chained mutants and even incremental mutation testing.
```
[the remark this answers is not in the document]
Comment (Lms24) [reply data/blog/js-mutation-testing-our-sdks.md:66]: Playwright. I should make this more clear that I'm gonna touch on this later. thx!

File: data/blog/js-mutation-testing-our-sdks.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
+---
+title: 'Mutation Testing our JavaScript SDKs'
```
[the remark this answers is not in the document]
Comment (Lms24) [reply data/blog/js-mutation-testing-our-sdks.md:2]: Hmm I was rather thinking about  "Mutation-testing our JavaScript SDKs" (s.t. mutation-testing is the verb). Does this make sense?

[the remark this answers is not in the document]
Comment (nicohrubec) [reply data/blog/js-mutation-testing-our-sdks.md]: Sounds a bit off to me but maybe that's just me

**Review by nicohrubec — APPROVED** [review]

File: data/blog/js-mutation-testing-our-sdks.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+3. Repeat 1 and 2 until all possible mutant operators have been applied to your codebase or you reached a pre-defined maximum number of mutants.
+
+After this virtual bloodbath, you can calculate a _Mutation Score_ by dividing the number of killed mutants by the number of all created mutants. The score now tells you how likely it is that your tests would catch an actual bug.
+
+### Mutant Similarity Assumption
+
+The entire concept of MT depends on the assumption that [mutants are similar to actual bugs](https://dl.acm.org/doi/abs/10.1145/2635868.2635929) that programmers might introduce into the code base.
+Now the question is, is this true in reality? Luckily, engineers at Google publish papers about mutation testing every now and then. In [one of them](https://dl.acm.org/doi/10.1109/ICSE43902.2021.00087), they
+investigated the similarity assumption by comparing 15 million mutants against bugfix PRs. They found a high similarity of mutants to bugs, suggesting that the assumption does hold up.
+Does this translate to other code bases? Maybe not, but for the sake of trying things out, let's go with "yes".
```
Comment (mydea) [inline data/blog/js-mutation-testing-our-sdks.md:49]: Maybe you can re-phrase this, instead of saying "maybe not, but let's assume yes" you could write something like:

Does this translate to other code bases? It's hard to say with certainty, but it seems reasonable to assume that there is at least some level of relevance to most code bases.

File: data/blog/js-mutation-testing-our-sdks.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+### Performance
+
+You might have already guessed it but what really keeps mutation testing from being widely adopted is its high performance impact. Running 1000s of tests on 1000s of mutants is computationally expensive - much higher than any kind of coverage calculation. Fortunately, modern MT tools provide optimizations that minimize the runtime as much as possible. For example, they use "per-test" mutation coverage to determine which tests cover a mutant and then, only execute this subset of all tests on the specific mutant. Similarly, modern MT tools chain multiple mutants into one modified code base and activate a specific mutant at runtime. This means that the modified code base doesn't have to be re-built for every mutant, but instead only once at the beginning of the MT run.
```
Comment (mydea) [inline data/blog/js-mutation-testing-our-sdks.md:53]:
```suggestion
What really keeps mutation testing from being widely adopted is its high performance impact. Running 1000s of tests on 1000s of mutants is computationally expensive - much higher than any kind of coverage calculation. Fortunately, modern MT tools provide optimizations that minimize the runtime as much as possible. For example, they use "per-test" mutation coverage to determine which tests cover a mutant and then, only execute this subset of all tests on the specific mutant. Similarly, modern MT tools chain multiple mutants into one modified code base and activate a specific mutant at runtime. This means that the modified code base doesn't have to be re-built for every mutant, but instead only once at the beginning of the MT run.
```

not sure if people might have guessed this - best to avoid assumptions in writing ;)

Comment (Lms24) [reply data/blog/js-mutation-testing-our-sdks.md:53]: Thanks for the feedback, agree with both items! I opened https://github.com/getsentry/sentry.engineering/pull/142 with the rewordings.
